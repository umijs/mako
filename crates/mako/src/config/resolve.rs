use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug)]
pub struct ResolveConfig {
    pub alias: Vec<(String, String)>,
    pub extensions: Vec<String>,
}
