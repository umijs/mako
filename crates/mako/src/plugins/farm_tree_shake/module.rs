use std::collections::{HashMap, HashSet};

use mako_core::swc_ecma_ast::{Module as SwcModule, ModuleItem};

use crate::module::{Module, ModuleId};
use crate::plugins::farm_tree_shake::statement_graph::{
    ExportInfo, ExportInfoMatch, ExportSpecifierInfo, ImportInfo, StatementGraph, StatementId,
};
use crate::tree_shaking::tree_shaking_module::ModuleSystem;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum UsedIdent {
    /// Local ident
    SwcIdent(String),
    /// Default ident
    Default,
    /// This ident is used and may be exported from other module
    InExportAll(String),
    /// All idents is used and may be exported from other module
    ExportAll,
}

impl ToString for UsedIdent {
    fn to_string(&self) -> String {
        match self {
            UsedIdent::SwcIdent(ident) => ident.to_string(),
            UsedIdent::Default => "default".to_string(),
            UsedIdent::InExportAll(ident) => ident.to_string(),
            UsedIdent::ExportAll => "*".to_string(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum UsedExports {
    All,
    Partial(HashSet<String>),
}

impl UsedExports {
    pub fn add_used_export(&mut self, used_export: &dyn ToString) -> bool {
        match self {
            UsedExports::All => {
                *self = UsedExports::All;
                false
            }
            UsedExports::Partial(self_used_exports) => {
                self_used_exports.insert(used_export.to_string())
            }
        }
    }

    pub fn use_all(&mut self) -> bool {
        match self {
            UsedExports::All => false,
            UsedExports::Partial(_) => {
                *self = UsedExports::All;
                // fixme : case partial used all exports
                true
            }
        }
    }

    pub fn is_empty(&self) -> bool {
        match self {
            UsedExports::All => false,
            UsedExports::Partial(self_used_exports) => self_used_exports.is_empty(),
        }
    }
}

#[derive(Clone, Debug)]
pub enum AllExports {
    Precise(HashSet<String>),
    Ambiguous(HashSet<String>),
}

impl AllExports {
    fn all_specifiers(&self) -> Vec<String> {
        match self {
            AllExports::Precise(ids) => ids
                .iter()
                .cloned()
                .filter(|id| id != "default")
                .collect::<Vec<_>>(),

            AllExports::Ambiguous(ids) => ids
                .iter()
                .cloned()
                .filter(|id| id != "default")
                .collect::<Vec<_>>(),
        }
    }

    pub fn to_all_specifier(&self) -> ExportSpecifierInfo {
        let sps = self.all_specifiers();

        match self {
            AllExports::Precise(_) => ExportSpecifierInfo::All(sps),
            AllExports::Ambiguous(_) => ExportSpecifierInfo::Ambiguous(sps),
        }
    }

    pub fn add_idents<I: IntoIterator<Item = String>>(&mut self, idents: I) {
        match self {
            AllExports::Precise(s) => s.extend(idents),
            AllExports::Ambiguous(s) => s.extend(idents),
        }
    }
}

pub struct TreeShakeModule {
    pub module_id: ModuleId,
    pub side_effects: bool,
    pub stmt_graph: StatementGraph,
    // used exports will be analyzed when tree shaking
    pub used_exports: UsedExports,
    pub module_system: ModuleSystem,
    pub all_exports: AllExports,
    pub topo_order: usize,
    pub updated_ast: Option<SwcModule>,
}

impl TreeShakeModule {
    pub fn extends_exports(&mut self, to_extend: &AllExports) {
        match (&mut self.all_exports, to_extend) {
            (AllExports::Precise(me), AllExports::Precise(to_add)) => {
                me.extend(to_add.iter().cloned());
            }
            (AllExports::Ambiguous(me), AllExports::Precise(to_add)) => {
                me.extend(to_add.iter().cloned());
            }
            (AllExports::Precise(me), AllExports::Ambiguous(to_add)) => {
                me.extend(to_add.iter().cloned());

                self.all_exports = AllExports::Ambiguous(me.clone())
            }
            (AllExports::Ambiguous(me), AllExports::Ambiguous(to_add)) => {
                me.extend(to_add.iter().cloned());
            }
        }
    }

    pub fn new(module: &Module, order: usize) -> Self {
        let module_info = module.info.as_ref().unwrap();

        // 1. generate statement graph
        let mut module_system = ModuleSystem::CommonJS;
        let stmt_graph = match &module_info.ast {
            crate::module::ModuleAst::Script(module) => {
                let is_esm = module
                    .ast
                    .body
                    .iter()
                    .any(|s| matches!(s, ModuleItem::ModuleDecl(_)));
                if is_esm {
                    module_system = ModuleSystem::ESModule;
                    StatementGraph::new(&module.ast)
                } else {
                    StatementGraph::empty()
                }
            }
            crate::module::ModuleAst::Css(_) => {
                module_system = ModuleSystem::Custom;
                StatementGraph::empty()
            }
            crate::module::ModuleAst::None => {
                module_system = ModuleSystem::Custom;
                StatementGraph::empty()
            }
        };

        // 2. set default used exports
        let used_exports = if module.side_effects {
            UsedExports::All
        } else {
            UsedExports::Partial(Default::default())
        };

        Self {
            module_id: module.id.clone(),
            stmt_graph,
            used_exports,
            side_effects: module.side_effects,
            all_exports: match module_system {
                ModuleSystem::ESModule => AllExports::Precise(Default::default()),
                ModuleSystem::Custom | ModuleSystem::CommonJS => {
                    AllExports::Ambiguous(Default::default())
                }
            },
            module_system,
            topo_order: order,
            updated_ast: None,
        }
    }

    pub fn imports(&self) -> Vec<ImportInfo> {
        let mut imports = vec![];

        for stmt in self.stmt_graph.stmts() {
            if let Some(import) = &stmt.import_info {
                imports.push(import.clone());
            }
        }

        imports
    }

    pub fn exports(&self) -> Vec<ExportInfo> {
        let mut exports = vec![];

        for stmt in self.stmt_graph.stmts() {
            if let Some(export) = &stmt.export_info {
                exports.push(export.clone());
            }
        }

        exports
    }

    pub fn used_statements(&self) -> HashMap<StatementId, HashSet<String>> {
        // 1. get used exports
        let used_exports_idents = self.used_exports_idents();
        let mut stmt_used_idents_map = HashMap::new();

        for (used_ident, stmt_id) in used_exports_idents {
            let used_idents: &mut HashSet<UsedIdent> =
                stmt_used_idents_map.entry(stmt_id).or_default();
            used_idents.insert(used_ident);
        }

        {
            for stmt in self.stmt_graph.stmts() {
                if stmt.is_self_executed {
                    stmt_used_idents_map.entry(stmt.id).or_default();

                    let dep_stmts = self.stmt_graph.dependencies(&stmt.id);

                    for (dep_stmt, referred_idents) in dep_stmts {
                        let used_idents = stmt_used_idents_map.entry(dep_stmt.id).or_default();
                        used_idents.extend(referred_idents.into_iter().map(UsedIdent::SwcIdent));
                    }
                    // stmt.used_idents.iter().for_each(|used_ident| {
                    //   // find the defined ident
                    //   stmt_used_idents_map
                    //   .entry(stmt.id)
                    //   .or_insert(HashSet::new());

                    //   for stmt_inner in self.stmt_graph.stmts() {
                    //     if stmt_inner.id == stmt.id {
                    //       continue;
                    //     }

                    //     if stmt_inner
                    //       .defined_idents_map
                    //       .contains_key(&used_ident.to_string())
                    //       || stmt_inner
                    //         .defined_idents
                    //         .iter()
                    //         .any(|ident| ident.to_string() == used_ident.to_string())
                    //     {
                    //       let used_idents = stmt_used_idents_map
                    //         .entry(stmt_inner.id)
                    //         .or_insert(HashSet::new());
                    //       used_idents.insert(UsedIdent::SwcIdent(used_ident.clone()));
                    //     }
                    //   }
                    // });
                }
            }
        }

        // 2. analyze used statements starting from used exports

        // println!("before {:?}", self.module_id);
        // dbg!(&used);
        //
        // for imp in self.imports() {
        //     if imp.source.contains("@swc/helpers") {
        //         used.entry(imp.stmt_id)
        //             .or_insert(HashSet::new());
        //     }
        // }
        //
        // dbg!(&used);

        self.stmt_graph
            .analyze_used_statements_and_idents(stmt_used_idents_map)
    }

    pub fn used_exports_idents(&self) -> Vec<(UsedIdent, StatementId)> {
        match &self.used_exports {
            UsedExports::All => {
                // all exported identifiers are used
                let mut used_idents = vec![];

                for export_info in self.exports() {
                    for sp in export_info.specifiers {
                        match sp {
                            ExportSpecifierInfo::Default => {
                                used_idents.push((UsedIdent::Default, export_info.stmt_id));
                            }
                            ExportSpecifierInfo::Named { local, .. } => {
                                used_idents.push((
                                    UsedIdent::SwcIdent(local.clone()),
                                    export_info.stmt_id,
                                ));
                            }
                            ExportSpecifierInfo::Namespace(ns) => {
                                used_idents
                                    .push((UsedIdent::SwcIdent(ns.clone()), export_info.stmt_id));
                            }
                            ExportSpecifierInfo::All(_) | ExportSpecifierInfo::Ambiguous(_) => {
                                used_idents.push((UsedIdent::ExportAll, export_info.stmt_id));
                            }
                        }
                    }
                }

                used_idents
            }
            // import {x,y,z} from "x"
            UsedExports::Partial(idents) => {
                let mut used_idents = vec![];

                for ident in idents {
                    // find the export info*s* that contains the ident

                    let mut export_infos = vec![];

                    for export_info in self.exports().into_iter() {
                        match export_info.matches_ident(ident) {
                            ExportInfoMatch::Matched => {
                                export_infos = vec![export_info];
                                break;
                            }
                            ExportInfoMatch::Unmatched => {}
                            ExportInfoMatch::Ambiguous => {
                                export_infos.push(export_info);
                            }
                        }
                    }

                    for export_info in export_infos {
                        for sp in export_info.specifiers {
                            match sp {
                                ExportSpecifierInfo::Default => {
                                    if ident == "default" {
                                        used_idents.push((UsedIdent::Default, export_info.stmt_id));
                                    }
                                }
                                ExportSpecifierInfo::Named { local, exported } => {
                                    if let Some(exported) = exported {
                                        if is_ident_equal(ident, &exported) {
                                            used_idents.push((
                                                UsedIdent::SwcIdent(local.clone()),
                                                export_info.stmt_id,
                                            ));
                                        }
                                    } else if is_ident_equal(ident, &local) {
                                        used_idents.push((
                                            UsedIdent::SwcIdent(local.clone()),
                                            export_info.stmt_id,
                                        ));
                                    }
                                }
                                ExportSpecifierInfo::Namespace(ns) => {
                                    if is_ident_equal(ident, &ns) {
                                        used_idents.push((
                                            UsedIdent::SwcIdent(ns.clone()),
                                            export_info.stmt_id,
                                        ));
                                    }
                                }
                                ExportSpecifierInfo::All(exports_idents) => {
                                    let found = exports_idents.iter().find(|exported_ident| {
                                        is_ident_equal(ident, exported_ident)
                                    });

                                    if found.is_some() {
                                        used_idents.push((
                                            UsedIdent::InExportAll(ident.clone()),
                                            export_info.stmt_id,
                                        ));
                                    }
                                }
                                ExportSpecifierInfo::Ambiguous(_) => used_idents.push((
                                    UsedIdent::InExportAll(ident.clone()),
                                    export_info.stmt_id,
                                )),
                            }
                        }
                    }
                }

                used_idents
            }
        }
    }
}

pub fn is_ident_equal(ident1: &str, ident2: &str) -> bool {
    let split1 = ident1.split('#').collect::<Vec<_>>();
    let split2 = ident2.split('#').collect::<Vec<_>>();

    if split1.len() == 2 && split2.len() == 2 {
        split1[0] == split2[0] && split1[1] == split2[1]
    } else {
        split1[0] == split2[0]
    }
}
