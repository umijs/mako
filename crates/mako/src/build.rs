pub(crate) mod analyze_deps;
pub(crate) mod load;
pub(crate) mod parse;
pub(crate) mod targets;
pub(crate) mod transform;

use std::collections::HashSet;
use std::sync::mpsc::channel;
use std::sync::Arc;

use anyhow::Result;
use colored::Colorize;
use thiserror::Error;

use crate::ast::file::{Content, File, JsContent};
use crate::ast::utils::get_module_system;
use crate::compiler::{Compiler, Context};
use crate::generate::chunk_pot::util::hash_hashmap;
use crate::module::{Module, ModuleAst, ModuleId, ModuleInfo};
use crate::plugin::NextBuildParam;
use crate::resolve::ResolverResource;
use crate::utils::thread_pool;

#[derive(Debug, Error)]
pub enum BuildError {
    #[error(
        "{:}\n{:}", "Build failed.".to_string().red().to_string(), errors.iter().map(| e | e.to_string()).collect::< Vec < _ >> ().join("\n")
    )]
    BuildTasksError { errors: Vec<anyhow::Error> },
}

impl Compiler {
    pub fn build(&self, files: Vec<File>) -> Result<HashSet<ModuleId>> {
        if files.is_empty() {
            return Ok(HashSet::new());
        }

        let (rs, rr) = channel::<Result<Module>>();

        let build_with_pool = |file: File, parent_resource: Option<ResolverResource>| {
            let rs = rs.clone();
            let context = self.context.clone();
            thread_pool::spawn(move || {
                let result = Self::build_module(&file, parent_resource, context.clone());
                let result = Self::handle_build_result(result, &file, context);
                rs.send(result).unwrap();
            });
        };
        let mut count = 0;
        for file in files {
            count += 1;
            build_with_pool(file, None);
        }

        let mut errors = vec![];
        let mut module_ids = HashSet::new();

        for build_result in rr {
            count -= 1;

            // handle build_module error
            if build_result.is_err() {
                errors.push(build_result.err().unwrap());
                if count == 0 {
                    break;
                } else {
                    continue;
                }
            }
            let module = build_result.unwrap();
            let module_id = module.id.clone();

            // update context.modules_with_missing_deps (watch only)
            if self.context.args.watch {
                if module.info.as_ref().unwrap().deps.missing_deps.is_empty() {
                    self.context
                        .modules_with_missing_deps
                        .write()
                        .unwrap()
                        .retain(|id| id != &module_id.id);
                } else {
                    self.context
                        .modules_with_missing_deps
                        .write()
                        .unwrap()
                        .push(module_id.id.clone());
                }
            }

            let mut module_graph = self.context.module_graph.write().unwrap();

            // handle current module
            let info = module.info.as_ref().unwrap();
            let resolved_deps = info.deps.resolved_deps.clone();
            let m = module_graph.get_module_mut(&module.id);
            if let Some(m) = m {
                m.set_info(module.info);
            } else {
                module_ids.insert(module.id.clone());
                module_graph.add_module(module);
            }

            // handle deps
            for dep in resolved_deps {
                let path = dep.resolver_resource.get_resolved_path();
                let dep_module_id = ModuleId::new(path.clone());
                if !module_graph.has_module(&dep_module_id) {
                    let module = match dep.resolver_resource {
                        ResolverResource::Virtual(_) | ResolverResource::Resolved(_) => {
                            let file = File::new(path.clone(), self.context.clone());

                            if self.context.plugin_driver.next_build(&NextBuildParam {
                                current_module: &module_id,
                                next_file: &file,
                                resource: &dep.resolver_resource,
                            }) {
                                count += 1;
                                build_with_pool(file, Some(dep.resolver_resource.clone()));
                            }

                            Self::create_empty_module(&dep_module_id)
                        }
                        ResolverResource::External(_) => Self::create_external_module(
                            &dep.resolver_resource,
                            self.context.clone(),
                        ),
                        ResolverResource::Ignored(_) => {
                            Self::create_ignored_module(&path, self.context.clone())
                        }
                    };

                    // 拿到依赖之后需要直接添加 module 到 module_graph 里，不能等依赖 build 完再添加
                    // 是因为由于是异步处理各个模块，后者会导致大量重复任务的 build_module 任务（3 倍左右）
                    module_ids.insert(module.id.clone());
                    module_graph.add_module(module);
                }
                module_graph.add_dependency(&module_id, &dep_module_id, dep.dependency);
            }
            if count == 0 {
                break;
            }
        }
        drop(rs);

        if !errors.is_empty() {
            return Err(anyhow::anyhow!(BuildError::BuildTasksError { errors }));
        }

        Ok(module_ids)
    }

    pub fn create_external_module(
        resolved_resource: &ResolverResource,
        context: Arc<Context>,
    ) -> Module {
        let external_name = resolved_resource
            .get_external()
            // safe
            .unwrap();
        let external_script = resolved_resource.get_script();
        let is_async = external_script.is_some();
        let origin_path = resolved_resource.get_resolved_path();
        let path = format!("virtual:external_{}", origin_path);
        let mut file = File::new(path.clone(), context.clone());
        let code = if let Some(url) = external_script {
            format!(
                r#"
module.exports = new Promise((resolve, reject) => {{
__mako_require__.loadScript('{}', (e) => e.type === 'load' ? resolve() : reject(e));
}}).then(() => {});
"#,
                url, external_name
            )
        } else {
            format!("module.exports = {};", external_name)
        };
        file.set_content(Content::Js(JsContent {
            content: code,
            ..Default::default()
        }));
        let ast = parse::Parse::parse(&file, context)
            // safe
            .unwrap();
        let raw = file.get_content_raw();
        let info = ModuleInfo {
            file,
            module_system: get_module_system(&ast),
            ast,
            external: Some(external_name),
            is_async,
            resolved_resource: Some(resolved_resource.clone()),
            raw,
            ..Default::default()
        };
        let module_id = ModuleId::new(origin_path.to_string());
        Module::new(module_id, false, Some(info))
    }

    fn create_error_module(file: &File, err: String, context: Arc<Context>) -> Result<Module> {
        let mut file = file.clone();
        let code = format!("throw new Error(`Module build failed:\n{:}`)", err);
        file.set_content(Content::Js(JsContent {
            content: code,
            ..Default::default()
        }));
        let ast = parse::Parse::parse(&file, context.clone())?;
        let path = file.path.to_string_lossy().to_string();
        let module_id = ModuleId::new(path.clone());
        let raw = file.get_content_raw();
        let info = ModuleInfo {
            file,
            module_system: get_module_system(&ast),
            ast,
            raw,
            ..Default::default()
        };
        Ok(Module::new(module_id, false, Some(info)))
    }

    fn create_ignored_module(path: &str, context: Arc<Context>) -> Module {
        let module_id = ModuleId::new(path.to_owned());

        let mut module = Module::new(module_id, false, None);

        let info = {
            let file = File::with_content(
                path.to_owned(),
                Content::Js(JsContent {
                    content: "export {};".to_string(),
                    ..Default::default()
                }),
                context.clone(),
            );
            let ast = parse::Parse::parse(&file, context.clone()).unwrap();

            ModuleInfo {
                file,
                module_system: get_module_system(&ast),
                ast,
                is_ignored: true,
                ..Default::default()
            }
        };

        module.set_info(Some(info));

        module
    }

    pub fn create_empty_module(module_id: &ModuleId) -> Module {
        Module::new(module_id.clone(), false, None)
    }

    pub fn handle_build_result(
        result: Result<Module>,
        file: &File,
        context: Arc<Context>,
    ) -> Result<Module> {
        if result.is_err() && context.args.watch {
            let module = Self::create_error_module(
                file,
                result.err().unwrap().to_string(),
                context.clone(),
            )?;
            Ok(module)
        } else {
            result
        }
    }

    pub fn build_module(
        file: &File,
        parent_resource: Option<ResolverResource>,
        context: Arc<Context>,
    ) -> Result<Module> {
        // 1. load
        let mut file = file.clone();
        let mut content = load::Load::load(&file, context.clone())?;
        let content = context.plugin_driver.load_transform(
            &mut content,
            &file.path.to_string_lossy(),
            &context,
        )?;
        file.set_content(content);

        // 2. parse
        let mut ast = parse::Parse::parse(&file, context.clone())?;

        // 3. transform
        transform::Transform::transform(&mut ast, &file, context.clone())?;

        // 4. analyze deps + resolve
        let deps = analyze_deps::AnalyzeDeps::analyze_deps(&ast, &file, context.clone())?;

        // 5. create module
        let path = file.path.to_string_lossy().to_string();
        let module_id = ModuleId::new(path.clone());
        let raw = file.get_content_raw();
        let is_entry = file.is_entry;
        let source_map_chain = file.get_source_map_chain(context.clone());
        let top_level_await = match &ast {
            ModuleAst::Script(ast) => ast.contains_top_level_await,
            _ => false,
        };
        let is_async_module = file.extname == "wasm";
        let is_async = is_async_module || top_level_await;

        // raw_hash is only used in watch mode
        // so we don't need to calculate when watch is off
        let raw_hash = if context.args.watch {
            file.get_raw_hash()
                .wrapping_add(hash_hashmap(&deps.missing_deps))
        } else {
            0
        };
        let info = ModuleInfo {
            file,
            deps,
            module_system: get_module_system(&ast),
            ast,
            resolved_resource: parent_resource,
            source_map_chain,
            top_level_await,
            is_async,
            raw_hash,
            raw,
            ..Default::default()
        };
        let module = Module::new(module_id, is_entry, Some(info));
        Ok(module)
    }
}
