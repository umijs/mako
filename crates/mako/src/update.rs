use std::collections::{HashMap, HashSet, VecDeque};
use std::fmt::{self, Error};
use std::path::PathBuf;
use std::sync::Arc;

use anyhow::{Ok, Result};
use rayon::prelude::*;
use tracing::debug;

use crate::build::Task;
use crate::compiler::Compiler;
use crate::module::{Dependency, Module, ModuleId};
use crate::resolve::{get_resolvers, Resolvers};
use crate::transform_in_generate::transform_modules;

#[allow(dead_code)]
pub enum UpdateType {
    Add,
    Remove,
    Modify,
}

#[derive(Default, Debug)]
pub struct UpdateResult {
    // 新增的模块Id
    pub added: HashSet<ModuleId>,
    // 删除的模块Id
    pub removed: HashSet<ModuleId>,
    // 修改的模块Id
    pub modified: HashSet<ModuleId>,
}

impl UpdateResult {
    pub fn is_updated(&self) -> bool {
        !self.modified.is_empty() || !self.added.is_empty() || !self.removed.is_empty()
    }
}

impl fmt::Display for UpdateResult {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut added = self.added.iter().map(|f| f.id.clone()).collect::<Vec<_>>();
        added.sort_by_key(|id| id.to_string());
        let mut modified = self
            .modified
            .iter()
            .map(|f| f.id.clone())
            .collect::<Vec<_>>();
        modified.sort_by_key(|id| id.to_string());
        let mut removed = self
            .removed
            .iter()
            .map(|f| f.id.clone())
            .collect::<Vec<_>>();
        removed.sort_by_key(|id| id.to_string());
        write!(
            f,
            r#"
added:{:?}
modified:{:?}
removed:{:?}
"#,
            &added, &modified, &removed
        )
    }
}

impl Compiler {
    pub fn update(&self, paths: Vec<(PathBuf, UpdateType)>) -> Result<UpdateResult> {
        let mut update_result: UpdateResult = Default::default();

        let resolvers = Arc::new(get_resolvers(&self.context.config));

        // watch 到变化的文件，如果不在在前的 module graph 中，需过滤掉
        let paths: Vec<(PathBuf, UpdateType)> = {
            let module_graph = self.context.module_graph.read().unwrap();
            paths
                .into_iter()
                .filter(|(p, _)| module_graph.has_module(&p.clone().into()))
                .collect()
        };

        // 先分组
        let mut modified = vec![];
        let mut removed = vec![];
        let mut added = vec![];
        for (path, update_type) in paths {
            match update_type {
                UpdateType::Add => {
                    added.push(path);
                }
                UpdateType::Remove => {
                    removed.push(path);
                }
                UpdateType::Modify => {
                    modified.push(path);
                }
            }
        }

        // 先做删除
        let removed_module_ids = self.build_by_remove(removed);
        update_result.removed.extend(removed_module_ids);

        // 分析修改的模块，结果中会包含新增的模块
        let (modified_module_ids, add_paths) = self
            .build_by_modify(modified, resolvers.clone())
            .map_err(|_| Error {})?;
        added.extend(add_paths);
        debug!("added:{:?}", &added);
        update_result.modified.extend(modified_module_ids);

        // 最后做添加
        let added_module_ids = self.build_by_add(&added, resolvers);
        update_result.added.extend(
            added
                .into_iter()
                .map(ModuleId::from_path)
                .collect::<HashSet<_>>(),
        );
        update_result.added.extend(added_module_ids);
        debug!("update_result:{:?}", &update_result);

        // 对有修改的模块执行一次 transform
        self.transform_for_change(&update_result)?;

        Result::Ok(update_result)
    }

    fn transform_for_change(&self, update_result: &UpdateResult) -> Result<()> {
        let mut changes: Vec<ModuleId> = vec![];
        for module_id in &update_result.added {
            changes.push(module_id.clone());
        }
        for module_id in &update_result.modified {
            changes.push(module_id.clone());
        }
        transform_modules(changes, &self.context)?;
        Ok(())
    }

    fn build_by_modify(
        &self,
        modified: Vec<PathBuf>,
        resolvers: Arc<Resolvers>,
    ) -> Result<(HashSet<ModuleId>, Vec<PathBuf>)> {
        let result = modified
            .par_iter()
            .map(|entry| {
                // first build
                let (module, dependencies, _) = Compiler::build_module(
                    self.context.clone(),
                    Task {
                        path: entry.to_string_lossy().to_string(),
                        is_entry: false,
                    },
                    resolvers.clone(),
                )?;

                // diff
                let module_graph = self.context.module_graph.read().unwrap();
                let current_dependencies: Vec<(ModuleId, Dependency)> = module_graph
                    .get_dependencies(&module.id)
                    .into_iter()
                    .map(|(module_id, dep)| (module_id.clone(), dep.clone()))
                    .collect();
                drop(module_graph);

                let mut add_modules: HashMap<ModuleId, Module> = HashMap::new();
                let mut target_dependencies: Vec<(ModuleId, Dependency)> = vec![];
                dependencies.into_iter().for_each(|(path, external, dep)| {
                    let module_id = ModuleId::new(path.clone());
                    // TODO: handle error
                    let module = self.create_module(external, path, &module_id).unwrap();
                    target_dependencies.push((module_id.clone(), dep));
                    add_modules.insert(module_id, module);
                });

                let d = diff(current_dependencies, target_dependencies);
                Result::Ok((module, d.added, d.removed, add_modules))
            })
            .collect::<Result<Vec<_>>>();
        let result = result?;

        let mut added = vec![];
        let mut modified_module_ids = HashSet::new();

        let mut module_graph = self.context.module_graph.write().unwrap();
        for (module, add, remove, mut add_modules) in result {
            // remove bind dependency
            for (remove_module_id, _) in remove {
                module_graph.remove_dependency(&module.id, &remove_module_id)
            }

            // add bind dependency
            for (add_module_id, dep) in &add {
                let add_module = add_modules.remove(add_module_id).unwrap();

                // 只针对非 external 的模块设置 add Task
                if add_module.info.is_none() {
                    added.push(add_module_id.to_path());
                }

                module_graph.add_module(add_module);
                module_graph.add_dependency(&module.id, add_module_id, dep.clone());
            }

            modified_module_ids.insert(module.id.clone());

            // replace module
            module_graph.replace_module(module);
        }

        Result::Ok((modified_module_ids, added))
    }

    fn build_by_add(&self, added: &Vec<PathBuf>, resolvers: Arc<Resolvers>) -> HashSet<ModuleId> {
        let mut add_queue: VecDeque<Task> = VecDeque::new();
        for path in added {
            add_queue.push_back(Task {
                path: path.to_string_lossy().to_string(),
                is_entry: false,
            })
        }

        self.build_module_graph_by_task_queue(&mut add_queue, resolvers)
    }

    fn build_by_remove(&self, removed: Vec<PathBuf>) -> HashSet<ModuleId> {
        let mut removed_module_ids = HashSet::new();
        for path in removed {
            let from_module_id = ModuleId::from_path(path);
            let mut deps_module_ids = vec![];
            let mut module_graph = self.context.module_graph.write().unwrap();
            module_graph
                .get_dependencies(&from_module_id)
                .into_iter()
                .for_each(|(module_id, _)| {
                    deps_module_ids.push(module_id.clone());
                });
            for to_module_id in deps_module_ids {
                module_graph.remove_dependency(&from_module_id, &to_module_id);
            }
            module_graph.remove_module(&from_module_id);
            removed_module_ids.insert(from_module_id);
        }
        removed_module_ids
    }
}

pub struct Diff {
    added: HashSet<(ModuleId, Dependency)>,
    removed: HashSet<(ModuleId, Dependency)>,
}

// 对比两颗 Dependency 的差异
fn diff(right: Vec<(ModuleId, Dependency)>, left: Vec<(ModuleId, Dependency)>) -> Diff {
    let right: HashSet<(ModuleId, Dependency)> = right.into_iter().collect();
    let left: HashSet<(ModuleId, Dependency)> = left.into_iter().collect();
    let removed = right
        .difference(&left)
        .collect::<HashSet<_>>()
        .into_iter()
        .map(|dep| (*dep).clone())
        .collect();
    let added = left
        .difference(&right)
        .collect::<HashSet<_>>()
        .into_iter()
        .map(|dep| (*dep).clone())
        .collect();
    Diff { added, removed }
}

#[cfg(test)]
mod tests {

    use crate::module::ModuleId;
    use crate::test_helper::{module_to_jscode, setup_compiler, setup_files};
    use crate::update::UpdateType;
    use crate::{assert_debug_snapshot, assert_display_snapshot};

    #[tokio::test(flavor = "multi_thread")]
    async fn test_build() {
        let compiler = setup_compiler("test/build/tmp/single", true);
        setup_files(
            &compiler,
            vec![
                (
                    "index.ts".into(),
                    r#"
(async () => {
    await import('./chunk-1.ts');
})();
    "#
                    .into(),
                ),
                (
                    "chunk-1.ts".into(),
                    r#"
export default async function () {
    console.log(123);
}
    "#
                    .into(),
                ),
            ],
        );
        compiler.compile();
        {
            let module_graph = compiler.context.module_graph.read().unwrap();
            assert_display_snapshot!(&module_graph);
        }
        setup_files(
            &compiler,
            vec![
                (
                    "index.ts".into(),
                    r#"
(async () => {
    await import('./chunk-2.ts');
})();
"#
                    .into(),
                ),
                (
                    "chunk-2.ts".into(),
                    r#"
export const foo = 1;
"#
                    .into(),
                ),
            ],
        );
        let result = compiler
            .update(vec![(
                compiler.context.root.join("index.ts"),
                UpdateType::Modify,
            )])
            .unwrap();

        assert_display_snapshot!(&result);

        {
            let module_graph = compiler.context.module_graph.read().unwrap();
            assert_display_snapshot!(&module_graph);
        }
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_update_multi() {
        let compiler = setup_compiler("test/build/tmp/multi", true);
        let target_path = compiler.context.root.join("index.ts");
        setup_files(
            &compiler,
            vec![
                (
                    "index.ts".into(),
                    r#"
(async () => {
    await import('./chunk-1.ts');
})();
    "#
                    .into(),
                ),
                (
                    "chunk-1.ts".into(),
                    r#"
export default async function () {
    console.log(123);
}
    "#
                    .into(),
                ),
            ],
        );
        compiler.compile();
        {
            let module_graph = compiler.context.module_graph.read().unwrap();
            let code = module_to_jscode(&compiler, &ModuleId::from_path(target_path.clone()));
            assert_display_snapshot!(&module_graph);
            assert_debug_snapshot!(&code);
        }
        setup_files(
            &compiler,
            vec![
                (
                    "index.ts".into(),
                    r#"
(async () => {
    await import('./chunk-2.ts');
})();
"#
                    .into(),
                ),
                (
                    "chunk-2.ts".into(),
                    r#"
export * from './chunk-3.ts';
"#
                    .into(),
                ),
                (
                    "chunk-3.ts".into(),
                    r#"
export const foo = 1;
"#
                    .into(),
                ),
            ],
        );
        let result = compiler
            .update(vec![(target_path.clone(), UpdateType::Modify)])
            .unwrap();

        assert_display_snapshot!(&result);
        {
            let module_graph = compiler.context.module_graph.read().unwrap();
            let code = module_to_jscode(&compiler, &ModuleId::from_path(target_path));
            assert_display_snapshot!(&module_graph);
            assert_debug_snapshot!(&code);
        }
        {
            let module_graph = compiler.context.module_graph.read().unwrap();
            assert_display_snapshot!(&module_graph);
        }
    }
}
