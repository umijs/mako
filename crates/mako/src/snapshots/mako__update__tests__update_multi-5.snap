---
source: crates/mako/src/update.rs
expression: "&code"
---
"\"use strict\";\n(async ()=>{\n    await require.ensure([\n        \"<CWD>/crates/mako/test/build/tmp/multi/chunk-2.ts\"\n    ]).then(require.bind(require, \"<CWD>/crates/mako/test/build/tmp/multi/chunk-2.ts\"));\n})();\n\n//# sourceMappingURL=<CWD>/crates/mako/test/build/tmp/multi/index.ts.map"
